### PR TITLE
bear: fix

### PR DESCRIPTION
**Pull request by emilazy**
###### Description of changes

The previous build system patch was incomplete and left the `wrapper.d` directory empty, leaving Bear unable to pick up any build commands at all and breaking the functional test suite, which we also weren't running. Switch to overriding CMake flags instead and add the patches and dependencies necessary to get the functional tests running to prevent a future regression.

I've checked that the instructions in Nix's `doc/manual/src/contributing/hacking.md` work after this change, which is what started this yak shave in the first place.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
Adding backport labels back to 22.11, because that's [the version that Nix's flake pins to](https://github.com/NixOS/nix/blob/8f6255b86ac9c29bd96744300e5e26f989e1e2d5/flake.nix#L4) and it's awkward for Nix's hacking instructions to not work because of a package bug.

**Review by Atemu — APPROVED** [review]
LGTM

**Comment by emilazy** [comment]
Looks like some of the `LD_PRELOAD` tests fail on Linux. I will try to take a look. My guess is they're failing because of Nix wrappers adding additional unexpected `exec` calls or similar.

**Comment by emilazy** [comment]
I fixed the majority of the Linux test case failures but there are a few remaining ones relating to arcane LD_PRELOAD interception stuff. I will try to see if I can get them fixed soon.

**Comment by zombiezen** [comment]
As a casual observer who has also encountered the problems with the `wrapper.d` directory being empty, I hacked out [a gist](https://gist.github.com/zombiezen/07bb2bc32cd065dc00a2adf93012cff5) based on this PR.

My only real contribution here is that for cross-compiling, the tests should probably be separated out from the build phase. Bear doesn't make it easy though, and I don't think I even got it working right, but for my own purposes, I don't care about the tests passing. Hope this is helpful.

File: pkgs/development/tools/build-managers/bear/default.nix [hunk 1]
```diff
@@ -40,11 +52,29 @@ stdenv.mkDerivation rec {
...
+  postPatch = ''
+    patchShebangs test/bin
+
+    # /usr/bin/env is used in test commands and embedded scripts.
+    find test -name '*.sh' \
+      -exec sed -ie 's|/usr/bin/env|${coreutils}/bin/env|g' {} +
+  '';
+
+  # Functional tests use loopback networking.
+  __darwinAllowLocalNetworking = true;
```
Comment (AndersonTorres) [inline pkgs/development/tools/build-managers/bear/default.nix:76]: Is this some env var?

Comment (emilazy) [reply pkgs/development/tools/build-managers/bear/default.nix:76]: It is a Nix-handled primitive derivation attribute: https://github.com/NixOS/nix/blob/f878b422b0b245d62593a909173b10b2b0588eaf/src/libstore/build/local-derivation-goal.cc#L2110; there are other uses in the Nixpkgs tree already.

Sorry for dropping the ball on this PR; I have been busy with moving lately and will endeavour to get back to it (and my other Nix work) in the next few weeks.

**Comment by AndersonTorres** [comment]
@emilazy ping

**Comment by DieracDelta** [comment]
Any chance we could merge this, and fix any remaining tests later? This version of bear works significantly better for me than what's currently in nixpkgs.

**Comment by Atemu** [comment]
This needs a rebase as `bear` has been marked as broken on Darwin since.

**Comment by DieracDelta** [comment]
How can we move this forward? I'm happy do the rebase but I don't have perms to push to @emilazy 's branch.

**Comment by Atemu** [comment]
Go right ahead, I don't think @emilazy would mind.

**Comment by SuperSandro2000** [comment]
Closing in favor of #279035

**Pull request closed without merging**